README.md
[![Logo](http://i.imgur.com/fpVkTZk.png)](http://www.openmrs.org)

OpenMRS Android Client and SDK
==============================

[![Build Status Travis](https://travis-ci.org/openmrs/openmrs-contrib-android-client.svg?branch=master)](https://travis-ci.org/openmrs/openmrs-contrib-android-client) [![Build Status AppVeyor](https://ci.appveyor.com/api/projects/status/github/openmrs/openmrs-contrib-android-client?branch=master&svg=true)](https://ci.appveyor.com/project/AvijitGhosh82/openmrs-contrib-android-client) [![Demo Server](https://img.shields.io/badge/demo-online-green.svg)](http://devtest04.openmrs.org:8080/openmrs) [![GitHub version](https://d25lcipzij17d.cloudfront.net/badge.svg?id=gh&type=6&v=2.8.1&x2=0)](https://github.com/openmrs/openmrs-contrib-android-client/releases/latest) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/37fa8e86a3cb4256a3b7ffcc644f13c6)](https://www.codacy.com/app/marzeion-tomasz/openmrs-contrib-android-client?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=openmrs/openmrs-contrib-android-client&amp;utm_campaign=Badge_Grade) [![codecov](https://codecov.io/gh/openmrs/openmrs-contrib-android-client/branch/master/graph/badge.svg)](https://codecov.io/gh/f4ww4z/openmrs-contrib-android-client) [![IRC](https://img.shields.io/badge/IRC-%23openmrs-1e72ff.svg?style=flat)](http://irc.openmrs.org)

## Table of Contents
- [OpenMRS Android Client](#OpenMRS-Android-Client)
   - [Table of Contents](#Table-of-Contents)
- [Description](#Description)
         - [Key Features](#Key-Features)
- [Screenshots](#Screenshots)
- [Demo Server](#Demo-Server)
            - [Demo Username: admin](#Demo-Username-admin)
            - [Demo Password: Admin123](#Demo-Password-Admin123)
- [Releasing [Collaborators only]](#Releasing-Collaborators-only)
- [License](#License)
- [Resources](#Resources)

# Description
* The purpose of this project is to provide an OpenMRS client for Android devices. The app is designed to cover most of the functionality currently on the web application.
The app communicates with OpenMRS instances using REST. It supports working offline (without a network connection). The database on the device is encrypted and password-protected to secure patient data.
* In addition to the OpenMRS-Android-Client, this repository also has openmrs-android-sdk package which serves as an sdk which helps to build custom UI on top of the core functionality it provides.

# Development
* We use JIRA to track issues and monitor project development. Refer to this link to view all issues and project summary: [Android Client JIRA](https://issues.openmrs.org/browse/AC). 
To get started contributing, try working on [introductory issues](https://issues.openmrs.org/issues/?filter=17165) in JIRA and check out [OpenMRS Pull Request Tips](https://wiki.openmrs.org/display/docs/Pull+Request+Tips).
Also, before creating a pull request, please run code review tools (Lint) and all tests.

* There is a detailed guide for setting up the OpenMRS-Android-Client locally, before starting to contribute to the project [here](CONTRIBUTING.md).

#### 1. openmrs-android-sdk package
* There was a need to make the app extendable without forking it out, instead just adding it as a dependency in any android application which wants to use the functionality but with a custom UI on top of it.
* We can add the implementation    
``` 
    dependencies 
    {
        implementation 'com.github.openmrs:openmrs-contrib-android-client:deploy-android-sdk-SNAPSHOT'
    } 
```
as a dependency in the app module build.gradle to get the functionality provided by openmrs-android-sdk.
* The openmrs-android-sdk exposes the functionality through, methods present in various repository classes which just need to be plugged in with the UI and view-Model.
* An simple example of the usage would be [this demo application](https://github.com/LuGO0/Test-Application), a more complex application depicting the usage would be the openmrs-client package itself. There is a confluence article [here](
https://wiki.openmrs.org/display/docs/Getting+Started+with+openmrs-android-sdk) which will take you through the creation and usage of the test Application mentioned above.
* The JavaDocs for the openmrs-android-sdk can be generated by running the Gradle command ./gradle dokkaHtml which will generate an HTML interface with documentation inside the openmrs-android-sdk/build directory.

#### 2. openmrs-client package
* This package was earlier used as the sole package containing all the code for the OpenMRS-Android-Client now a part of it has been encapsulated in the form of openmrs-android-sdk and published on Jitpack from where it can simply be added as a dependency to any other app.
* This package uses the methods exposed by the openmrs-android-sdk and builds UI on top of it, which can be used as an example to implement UI on top of openmrs-android-sdk.
* The app is also published on PlayStore just to get used to the functionalities it provides and can be tested on local OpenMRS server or Demo OpenMRS Server.

#### Key Features
- Connect to OpenMRS server and sync all data
- Register and Edit patients
- Record Visits and Encounters
- View patient data (Details, Diagnoses, Visits, and Vitals)
- Allergies Module
- Provider Module
- Offline access (specific modules like provider, patient and some functionalities of allergy module)

# Screenshots
<img src="https://user-images.githubusercontent.com/45125121/82362785-804c2800-9a2a-11ea-9bb1-f1b778c70de5.jpg" width="280" height="520" alt="Login page" >  <img src="http://i.imgur.com/KmaWzNv.png" width="280" height="520"> <img src="http://i.imgur.com/hiCNNIx.png" width="280" height="520">
 
# Demo Server

The demo test server dedicated to the client is (https://demo.openmrs.org/openmrs/).
In case the demo server fails to respond, you can use other alternate servers provided [here.](https://wiki.openmrs.org/display/ISM/OpenMRS+environments)
##### Demo Username: admin
##### Password: Admin123
<br/>
 
 
 
# Releasing [Collaborators only]

### 1. OpenMRS-Android-Client to PlayStore
We follow the sprint model for development. Read more about it here: [OpenMRS Sprints](https://wiki.openmrs.org/display/RES/Development+Sprints).
To release the application, make sure to do these steps **in order**:

1. Update the [version variable in versions.gradle](https://github.com/openmrs/openmrs-contrib-android-client/blob/master/openmrs-client/versions.gradle#L6) prior to the release.
3. Update the [Release notes](releaseNotes.md) file.
4. Update the [release notes text file](https://github.com/openmrs/openmrs-contrib-android-client/blob/master/openmrs-client/src/main/play/release-notes/en-US/default.txt) to publish in the Play store. Ideally, change the wording so that normal end users understand.
5. Now commit with the title `Release <version number here>` to the master branch.
6. Tag the commit, using the version as the tag name. Make sure CI is green!
7. Go to [the releases page](https://github.com/openmrs/openmrs-contrib-android-client/releases) and click the [Draft a new release](https://github.com/openmrs/openmrs-contrib-android-client/releases/new) button. It will create a new version tag in the repository and build the app. The tag name will be used as the version number for this. Be sure to bump unfinished issues to the next due version.
8. Go to [JIRA's releases page](https://issues.openmrs.org/projects/AC?selectedItem=com.atlassian.jira.jira-projects-plugin:release-page), click on the three-dots on the right, and hit **Release**.
9. Post a new Talk thread and describe what is changed or improved in the release.

### 2. openmrs-android-sdk to jitpack
1. The Openmrs-Android-sdk gets published to the Jitpack library so that it can be added as a dependency in various projects.
2. Due to some issues with the release build configuration of the app is explained in detail [here](https://stackoverflow.com/questions/68420822/handling-release-keystore-while-uploading-android-library-to-jitpack) we are not able to do jitpack releases from the master branch.
3. There is a dedicated branch for this purpose that has got only the debug variant of the android application. So that the jitpack build passes.
4. For now until the issue gets resolved or we write a script for it, To release the latest code improvements in the openmrs-android-sdk package to the jitpack, we need to copy the whole package from master to branch deploy-android-sdk.
5. After getting the latest code changes to branch deploy-android-sdk, we publish a snapshot of the branch to jitpack. So that the tag of jitpack published looks exactly like `deploy-android-sdk-SNAPSHOT`.

# License
This project is licensed under the OpenMRS Public License, see the [copyright](copyright/copyright) file for details.

# Resources
- [User Guide](https://github.com/openmrs/openmrs-android-client-user-guide)
- [Contribution Guidelines](https://github.com/openmrs/openmrs-contrib-android-client/blob/master/CONTRIBUTING.md)
- [JIRA](https://issues.openmrs.org/browse/AC/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel)
- [Sprint board](https://issues.openmrs.org/secure/RapidBoard.jspa?rapidView=60)
- [Dashboard](https://issues.openmrs.org/secure/Dashboard.jspa?selectPageId=12851)
- [CI](https://travis-ci.org/openmrs/openmrs-contrib-android-client)
- [Google Play](https://play.google.com/store/apps/details?id=org.openmrs.mobile)
- [Release Notes](releaseNotes.md)


releaseNotes.md

# Release Notes

### Version 3.1.1

1. [AC-1068] - Not able to Login due to session id
2. [AC-1069] - Offline Patient click crashes the app
3. [AC-1070] - Identifier not created for the offline patient
4. [AC-1071] - Add new functions and enhancements in the SDK
5. [AC-1072] - Add more APIs for the Encounter operations in the SDK
6. [AC-1073] - Migrate EncounterCreateRoomDAO to Kotllin
7. [AC-1074] - Migrate ObservationRoomDAO to Kotlin
8. [AC-1075] - CI/CD pipeline fix for GitHub
9. [AC-1077] - Add/Enhance APIs for Observation operations in the SDK
10. [AC-1078] - Migrate VisitRoomDAO to Kotlin
11. [AC-1079] - Add/Enhance APIs for Visit operations in the SDK
12. [AC-1080] - Add Appointment Scheduling feature to the SDK
13. [AC-1081] - Migrate ConceptRoomDAO to Kotlin
14. [AC-1082] - Add Orders feature to the SDK
15. [AC-1088] - Add unit tests for the Appointment Repository
16. [AC-1089] - Add base changes for Unit testing
17. [AC-1090] - Add unit tests for Visit Repository
18. [AC-1091] - Add unit tests for Encounter Repository
19. [AC-1092] - Add unit tests for Observation Repository
20. [AC-1093] - Add Drugs Feature to the SDK
21. [AC-1094] - Add Program, Program Workflow and Program Workflow states to the SDK
22. [AC-1095] - Add CI check for Unit Tests
23. [AC-1096] - Migrate EncounterDAO to Kotlin
24. [AC-1097] - Migrate EncounterRoomDAO to Kotllin
25. [AC-1098] - Migrate EncounterTypeRoomDAO to Kotlin
26. [AC-1099] - Migrate FormResourceDAO to Kotlin
27. [AC-1100] - Migrate LocationRoomDAO to Kotllin
28. [AC-1101] - Migrate LocationDAO to Kotlin
29. [AC-1102] - Migrate ObservationDAO to Kotlin
30. [AC-1103] - Add yml for Jitpack
31. [AC-1104] - Migrate PatientRoomDAO to Kotlin
32. [AC-1105] - Migrate ProviderRoomDAO to Kotlin
33. [AC-1106] - Migrate VisitDAO to Kotlin
34. [AC-1107] - Add Unit Tests for Drug Repository and fix bugs
35. [AC-1108] - Add Unit Tests for Program Repository and fix bugs
36. [AC-1109] - Fix erros due to Kotlin Migration

### Version 3.0.0

1. [AC-574] - Android Client cannot update provider password
2. [AC-575] - Password Reset
3. [AC-584] - Renew GITHUB API key to deploy to Play store
4. [AC-657] - Adding an about section
5. [AC-717] - updating the PR template so that someone dosent forgets the JIRA workflow
6. [AC-722] - after starting a Visit replace the start button with an end button
7. [AC-821] - Fix deprecation warnings
8. [AC-869] - Full name of languages should be used.
9. [AC-906] - Re-design Home screen with a better UI.
10. [AC-926] - Using View Binding Architecture component in various packages.
11. [AC-945] - android.viewBinding.enabled is obsolete and should be replaced
12. [AC-951] - Migrate matchingpatients package from Java to Kotlin & ViewBinding
13. [AC-958] - Correcting app screenshots' aspect ratio in README.md file
14. [AC-962] - Stay on the SyncedPatientsActivity after deleting synced patient(s)
15. [AC-973] - Feature of setting system default theme
16. [AC-984] - Delete Operation Offline support for Allergy Repository
17. [AC-985] - Offline Strategy for Patient Repository
18. [AC-543] - Delay while inflating layout of Find Patients and Form Entry
19. [AC-545] - Loss of Data after Logout and re-login with same username,password & url
20. [AC-581] - Failed to resolve variable '${project.groupId}'
21. [AC-613] - Job 'generateReleasePlayResources' failed
22. [AC-616] - Encrypt play store files earlier in CI build
23. [AC-669] - Implementing a method to show detailed server error
24. [AC-774] - Travis CI fails to detect failed tests
25. [AC-775] - Theme bug causing glitch when opening register patient
26. [AC-792] - The Intent is not implemented properly
27. [AC-800] - Redesign register patient form as a multi step form
28. [AC-862] - On tapping email us in contact page, the recipient of mail is not correct
29. [AC-864] - The GetProviders method seems incorrect in the providerRepository
30. [AC-865] - We should avoid the use of appContext inside repository layer
31. [AC-866] - The title of action bars is not translated to Hindi.
32. [AC-883] - QuickFix for patientIdentifier Service omission
33. [AC-887] - Memory leak in Fragments
34. [AC-889] - An improvement is needed in the login page as in dark mode the username,password and especially the server url text colours make it not visible correctly.
35. [AC-890] - An improvement is needed in the Settings page UI. The language button English option text is not displayed correctly in some of the devices.
36. [AC-946] - Bug in the language switching feature.
37. [AC-961] - Buttons of alert dialog are not clearly visible in dark mode.
38. [AC-965] - Form Entry TextView Overlap with Image
39. [AC-967] - App Crashes after registering patient
40. [AC-968] - Provider name is not correctly saved in server when switching from offline to online mode
41. [AC-969] - Add Provider Worker is called even after syncing provider when switching from offline to online mode
42. [AC-970] - Client fails to edit or delete a newly created Provider
43. [AC-983] - After Deleting Providers the parent screen still holds old data untill a refresh.
44. [AC-986] - The Provider fetch function duplicates data in the feed
45. [AC-982] - NPE in Patient Photo Activity
46. [AC-981] - Application Rehaul for the Library code encapsulation
47. [AC-798] - Add and view supervised patients for the current provider
48. [AC-834] - Migrate room Entity classes to Kotlin
49. [AC-842] - Add Repository Abstraction for AddEditPatient Package
50. [AC-879] - Migrate Form List Package to Kotlin
51. [AC-888] - Migrate Login Screen to Kotlin
52. [AC-892] - Migrate Add Provider Screen to Kotlin
53. [AC-898] - Convert Java classes of Model Package to Kotlin
54. [AC-929] - Provider Manager in the Dashboard Screen is redirecting to wrong activity
55. [AC-948] - Migrate Logs Screen from MVP Architecture to MVVM architecture
56. [AC-950] - Migrate Provider Dashboard Package from Java to Kotlin.
    
### Version 2.9.0


WebRTC/js/OpenTokChatter.js
function OpenTokChattr(targetElem, roomId,session, options){
  _this = this;
  this.targetElem = $(targetElem);
  this.roomId = roomId;
  this.messages = [];
  this.users = {};
  this.options = options;
  this.session = session;
  this.initialized = false;
  this.initOpenTok();
  this.templates = {};
  this.initChattrTemplates();
  this.targetElem.append(Handlebars.compile(this.templates.base));
  this.targetElem.find("#chattr #roomId").html(this.roomId); 
  $("#chatInput").keyup(_this.checkKeyPress);
  this.uiActions();
  // Every 10 seconds update the times for everyone
  setInterval(function () {
    $('.chat p').each(function () {
      var date = $(this).attr('data-date');
      var timeDiff = _this._timeDifference(new Date(date), new Date());
      $(this).attr('title',timeDiff);
    });
  }, 10000);
}
OpenTokChattr.prototype = {
  _this:this,
  constructor: OpenTokChattr,
  initChattrTemplates: function(){
    _this.templates.base = $('#chattrBaseTpl').html();
    _this.templates.chat = $('#chattrChatTpl').html();
    _this.templates.status = $('#chattrStatusTpl').html();
    _this.templates.newUser = $('#chattrNewUserTpl').html();
    _this.templates.userLeave = $('#chattrUserLeaveTpl').html();
    _this.templates.update = $('#chattrUpdateTpl').html();
    _this.templates.userList = $('#chattrUserListTpl').html();
    _this.templates.help = $('#chattrHelpTpl').html();
    _this.templates.nameExists = $('#chattrNameExistsTpl').html();
    _this.templates.focus = $('#chattrFocusTpl').html();
    _this.templates.unfocus = $('#chattrUnfocusTpl').html();
  },
  initOpenTok: function(){
    _this.session.on({
      sessionConnected: function(sessionConnectEvent){
        _this.setName(_this._defaultNickname(_this.session.connection.connectionId));
        setTimeout(function(){_this.initialized = true;}, 2000);
      },
      signal: function(signal){
        var signalData = JSON.parse(signal.data);
        switch(signal.type){
          case "signal:chat":
            _this.messages.push({"type": "chat", data: signalData});
            _this.printMessage({"type": "chat", data: signalData});
            
            break;
          case "signal:name":
            var oldName = _this.getNickname(signalData.from);
            var nameData = {"oldName": oldName, "newName": signalData.newName}; 
            _this.users[signalData.from] = signalData.newName;
            _this.printMessage({"type": "status", data:nameData});
            break;
          case "signal:help":
            _this.printMessage({"type": "help", data:signalData});
            break;
          case "signal:generalUpdate":
            _this.printMessage({"type": "generalUpdate", data:signalData});
            break;
          case "signal:selfUpdate":
            _this.printMessage({"type": "selfUpdate", data:signalData});
            break;
          case "signal:pastMessages":
            if(!_this.initialized){
              _this.messages = signalData.messages;
              _this.users = signalData.users;
              _this.printMessages();
              _this.initialized = true;
            }
            break;
          case "signal:focus":
            _this.printMessage({
              type: "focus",
              data: signalData
            });
            room.focus(signalData.connectionId);
            break;
          case "signal:unfocus":
            _this.printMessage({
              type: "unfocus",
              data: signalData
            });
            room.unfocus();
            break;
        }
      },
      connectionCreated: function(event){
        if(_this.initialized){
          var connectionId = event.connection.connectionId;
          if(!(connectionId in _this.users)){
            _this.users[connectionId] = _this._defaultNickname(connectionId);
          }
          _this.sendSignal("pastMessages", {"messages":_this.messages, "users": _this.users}, event.connection);
        }
        _this.printMessage({"type": "newUser", data: event.connection.connectionId});
      },
      connectionDestroyed: function(event){
        _this.printMessage({"type": "userLeave", data:{"from":event.connection.connectionId}});
        delete _this.users[event.connection.connectionId];
      },
    });
  },
  close: function(){
    if(this.options.closeable)
      return this.options.closeable;
    else
      $("#chattr").hide();
  },
  uiActions: function(){
    $(".inner-chat").animate({scrollTop: $(".inner-chat")[0].scrollHeight},1000);
    if(this.options){
      if(this.options.closeable){
      }
    }
  },
  signalError: function(error){
    if(error){
      console.log("signal error: " + error.reason);
    }
  },
  sendSignal:function(type, data, to){
    var signalData = {
      type: type
    };

    if (data) {
      signalData.data = JSON.stringify(data);
    } else {
      signalData.data = "{}";
    }
    if (to) {
     signalData.to=to;
    }

    _this.session.signal(signalData,_this.signalError);
  },
  setName:function(name){
    _this.users[_this.session.connection.connectionId]=name;
  },
  //iterate through messages in printMessages
  printMessages: function(){
    for(var i = 0; i<_this.messages.length; i++){
      _this.printMessage(_this.messages[i]);
    }
  },
  printMessage: function(msg){
    var data = msg.data;
    var html = "";
    var tmplData = {};
    switch(msg.type){
      case "chat":
        tmplData.time = this._timeDifference(new Date(data.date),new Date());
        tmplData.nickname=data.name+": ";
        tmplData.message=decodeURI(data.text).replace(/</g, '&lt;').replace(/>/g, '&gt;');
        tmplData.cls = _this.isMe(data.from)?"from-me":"from-others";
        _this.appendToMessages('chat', tmplData);
        break;
      case "status":
        tmplData.oldName = data.oldName;
        tmplData.newName = data.newName;
        _this.appendToMessages('status', tmplData);
        break;
      case "newUser":
        if(!_this.isMe(data.from)||!data){
          tmplData.nickname = _this.getNickname(data);
          _this.appendToMessages('newUser', tmplData);
        }
        break;
      case "userLeave":
        if(!_this.isMe(data.from)||!data.from){
          tmplData.nickname = _this.getNickname(data.from);
          _this.appendToMessages('userLeave', tmplData);
        }
        break;
      case "generalUpdate":
        tmplData.text = data.text; 
        _this.appendToMessages('update', tmplData);
        break;
      case "selfUpdate":
          tmplData.cls = data.cls;
        if(_this.isMe(data.from)){
          tmplData.text = data.text; 
          tmplData.cls = data.cls;
          _this.appendToMessages('update', tmplData);
        }
        break;
      case "focus":
        tmplData.nick = _this.getNickname(data.connectionId);
        _this.appendToMessages('focus', tmplData);
        break;
      case "unfocus":
        _this.appendToMessages('unfocus', tmplData);
        break;
    }
    $("#messages").append(html);
    $(".inner-chat").scrollTop($(".inner-chat")[0].scrollHeight)
  },
  appendToMessages: function(template, data){
    var tpl = Handlebars.compile(_this.templates[template]);
    $("#chattr .inner-chat ul#messages").append(tpl(data));
  },
  checkKeyPress: function(e){
    var code = (e.keyCode ? e.keyCode : e.which);
    if(code !== 13) {
     return;
    } 
    var text = $.trim($("#chatInput").val());
    if(text.length===0)
       return;
    var parts = text.split(" ");
    switch(parts[0]){
      case "/name":
      case "/nick":
        _this.sendChangeNameSignal(parts[1]);
        break;
      case "/help":
        _this.sendHelpSignal();
        break;
      case "/list":
        _this.sendListSignal();
        break;
      case "/focus":
        _this.sendFocusSignal();
        break;
      case "/unfocus":
        _this.sendUnfocusSignal();
        break;
      default:
        _this.sendChat(text);
    }
    $("#chatInput").val("");
  },
  sendHelpSignal: function(){
    _this.sendSelfUpdate(_this.templates.help, "help");
  },
  sendChat: function(msg){
    var date = new Date();
    var data = {name: _this.getNickname(_this.session.connection.connectionId), text: encodeURI(msg), date: date, from: _this.session.connection.connectionId};
    _this.sendSignal("chat", data);
  },
  sendGeneralUpdate: function(msg){
    _this.sendSignal("generalUpdate", {"text": msg});
  },
  sendSelfUpdate: function(msg,cls){
    var data = {from: _this.session.connection.connectionId, text: msg, cls: cls};
    _this.sendSignal("selfUpdate", data);
  },
  sendChangeNameSignal: function(newName){
    for(var k in _this.users){
      if(_this.users[k]===newName){
        var tpl = Handlebars.compile(_this.templates.nameExists);
        _this.sendSelfUpdate(tpl({newName: newName}));
        return;
      }
    }
    var data = {from: _this.session.connection.connectionId, newName: newName};
    _this.sendSignal("name", data);
  },
  signalUpdateUsers: function(){
    _this.sendSignal("updateUsers", _this.users);
  },
  sendListSignal: function(){
    var names = [];
    for(var k in _this.users){
      names.push(_this.users[k]);
    }
    var html = "";
    var data = {users:[]};
    for(var i = 0; i<names.length; i++){  
      var userData = {name: names[i]};
      if(i==names.length-1)
        userData.last = "last";
      data.users.push(userData);
    }
    var tpl = Handlebars.compile(_this.templates.userList);
    _this.sendSelfUpdate(tpl(data));
  },
  sendFocusSignal: function() {
    if (_this.initialized && room.initialized && room.myStream) {
      _this.sendSignal("focus", {
        connectionId: _this.session.connection.connectionId,
        streamId: room.myStream.streamId
      });
    }
  },
  sendUnfocusSignal: function() {
    if (_this.initialized && room.initialized) {
      _this.sendSignal("unfocus");
    }
  },
  getNickname: function(connectionId){
    return _this.users[connectionId] || _this._defaultNickname(connectionId);
  },
  isMe: function(connectionId){
    return connectionId===_this.session.connection.connectionId;
  },

  //Helper Methods

  _timeDifference: function(d1,d2){
    var seconds = (d2.getTime()-d1.getTime())/1000;
    if(seconds>=60 && seconds<120)
      return "1 minute ago";
    else if(seconds>=120 && seconds<3600)
      return parseInt(seconds/60,10)+" minutes ago";
    else if(seconds>=3600 && seconds<7200)
      return "1 hour ago";
    else if (seconds>=7200)
      return parseInt(seconds/3600,10)+" hours ago";
    else if (seconds>=10)
      return parseInt(seconds/60,10)+" seconds ago";
    else
      return "Just now";
  },
  _defaultNickname: function(connectionId){
    return "Guest-"+connectionId.substring( connectionId.length - 8, connectionId.length )
  }
}
